Adds secilc and selinux-policy packages

doverride opened the pull request:
Package manifests for selinux-policy and secilc dependency

Cc. aparcar
Cc. dangowrt

kcinimod-defensec committed:
secilc: adds new package

Signed-off-by: Dominick Grift <dominick.grift@defensec.nl>

kcinimod-defensec committed:
selinux-policy: adds new package

Signed-off-by: Dominick Grift <dominick.grift@defensec.nl>

dangowrt commented:
Beautiful :1st_place_medal:
You are of course free to name it `dssp` or what ever you like, I guess @aparcar 's main concern was the fact that the word `openwrt` was part of the package name...

aparcar commented on package/utils/secilc/Makefile:
```diff
@@ -0,0 +1,65 @@
+endef
+
+define Build/Compile
+	$(call Build/Compile/Default,secilc)
```
If you just use the default you can remove these steps.

doverride replied:
Can't do that as that will add manual pages and pull in xmlto

aparcar commented on package/system/selinux-policy/Makefile:
```diff
@@ -0,0 +1,53 @@
+
+define Package/selinux-policy/description
+	Basic SELinux Security Policy designed specifically for
+	OpenWRT and written in Common Intermediate Language.
```
It's OpenWrt not OpenWRT

doverride replied:
Thanks, addressed

aparcar commented on package/system/selinux-policy/Makefile:
```diff
@@ -0,0 +1,53 @@
+PKG_SOURCE_URL:=https://git.defensec.nl/selinux-policy.git
+PKG_SOURCE_DATE:=2020-09-30
+PKG_SOURCE_VERSION:=1629f471639ad25402671298ef38954a0bd739a8
+PKG_MIRROR_HASH:=skip
```
Please add a mirror hash, try `make package/selinux-policy/check V=s`

doverride replied:
Thanks, addressed

dangowrt committed:
config: add option for dssp selinux policy

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

doverride commented:
I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.

doverride commented:
Currently the fakeroot-script depends on `setfiles` (policycoreutils/host) but `CONFIG_TARGET_ROOTFS_SECURITY_LABELS=y` does not pull that in automatically (i think it should), so I added a dependency on that (similar to the refpolicy package)

Also bumped the version

dangowrt commented:
> I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.
Isn't it easier to have it 'enforcing' and then users may change to permissive mode using cmdline parameter?

dangowrt commented:
> Currently the fakeroot-script depends on `setfiles` (policycoreutils/host) but `CONFIG_TARGET_ROOTFS_SECURITY_LABELS=y` does not pull that in automatically (i think it should), so I added a dependency on that (similar to the refpolicy package)

doverride commented:
> > I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.
> 
> Isn't it easier to have it 'enforcing' and then users may change to permissive mode using cmdline parameter?

Yes this is not practical anyway. The mode from config is essentially useless because it is read from squashfs. The only way to change mode is via the kernel boot line (when youre lucky enough to have access to it) *and* only when the corresponding SELInux kernel config is enabled (which it is currently not, so essentially you'll pretty always boot into enforcing mode)

doverride commented:
> > > I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.
> > 
> > 
> > Isn't it easier to have it 'enforcing' and then users may change to permissive mode using cmdline parameter?
> 
> Yes this is not practical anyway. The mode from config is essentially useless because it is read from squashfs. The only way to change mode is via the kernel boot line (when youre lucky enough to have access to it) _and_ only when the corresponding SELInux kernel config is enabled (which it is currently not, so essentially you'll pretty always boot into enforcing mode)

IMHO Ideally, for now, we set the mode from config to permissive, `CONFIG_SECURITY_SELINUX_DEVELOP=y` and `CONFIG_SECURITY_SELINUX_BOOTPARAM=y` because there are many devices with many different properties (high risk of failure)

This way by default the system boots in permissive mode, and if one wants to boot in enforcing mode then `enforcing=1` can be appended to the kernel boot line in grub.

dangowrt commented:
> > > > I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.
> > > 
> > > 
> > > Isn't it easier to have it 'enforcing' and then users may change to permissive mode using cmdline parameter?
> > 
> > 
> > Yes this is not practical anyway. The mode from config is essentially useless because it is read from squashfs. The only way to change mode is via the kernel boot line (when youre lucky enough to have access to it) _and_ only when the corresponding SELInux kernel config is enabled (which it is currently not, so essentially you'll pretty always boot into enforcing mode)
> 
> IMHO Ideally, for now, we set the mode from config to permissive, `CONFIG_SECURITY_SELINUX_DEVELOP=y` and `CONFIG_SECURITY_SELINUX_BOOTPARAM=y` because there are many devices with many different properties (high risk of failure)
> 
> This way by default the system boots in permissive mode, and if one wants to boot in enforcing mode then `enforcing=1` can be appended to the kernel boot line in grub.

There are no plans of enabling SELinux by default in OpenWrt in the near future.
However, we may ship an SELinux-enabled ImageBuilder to allow users to experiment with SELinux on platforms which have sufficient resources for that (mostly flash space, obviously). The SELinux-enabled IB would ship with an SELinux-enabled Kernel and also have `CONFIG_SECURITY_SELINUX_DEVELOP` and `CONFIG_SECURITY_SELINUX_BOOTPARAM` set.
Though this is not very important at that stage, I still believe it'd be better to set 'enforcing' in /etc/selinux/config, so the IB is suitable to build enforcing images (which is otherwise quite difficult and will require individually fiddling with bootloader parameters on each and every device, which cannot even be done persistently on many devices as they are lacking a read/write U-Boot environment).
On the other side, temporarily setting `enforcing=0` on the kernel cmdline for development purposes is acceptable imho.

doverride commented:
> > > > > I also set the default global mode to permissive for now in `package/system/selinux-policy/files/selinux-config`.
> > > > 
> > > > 
> > > > Isn't it easier to have it 'enforcing' and then users may change to permissive mode using cmdline parameter?
> > > 
> > > 
> > > Yes this is not practical anyway. The mode from config is essentially useless because it is read from squashfs. The only way to change mode is via the kernel boot line (when youre lucky enough to have access to it) _and_ only when the corresponding SELInux kernel config is enabled (which it is currently not, so essentially you'll pretty always boot into enforcing mode)
> > 
> > 
> > IMHO Ideally, for now, we set the mode from config to permissive, `CONFIG_SECURITY_SELINUX_DEVELOP=y` and `CONFIG_SECURITY_SELINUX_BOOTPARAM=y` because there are many devices with many different properties (high risk of failure)
> > This way by default the system boots in permissive mode, and if one wants to boot in enforcing mode then `enforcing=1` can be appended to the kernel boot line in grub.
> 
> There are no plans of enabling SELinux by default in OpenWrt in the near future.
> However, we may ship an SELinux-enabled ImageBuilder to allow users to experiment with SELinux on platforms which have sufficient resources for that (mostly flash space, obviously). The SELinux-enabled IB would ship with an SELinux-enabled Kernel and also have `CONFIG_SECURITY_SELINUX_DEVELOP` and `CONFIG_SECURITY_SELINUX_BOOTPARAM` set.
> Though this is not very important at that stage, I still believe it'd be better to set 'enforcing' in /etc/selinux/config, so the IB is suitable to build enforcing images (which is otherwise quite difficult and will require individually fiddling with bootloader parameters on each and every device, which cannot even be done persistently on many devices as they are lacking a read/write U-Boot environment).
> On the other side, temporarily setting `enforcing=0` on the kernel cmdline for development purposes is acceptable imho.
> 
> Another idea may be to have the buildsystem generate /etc/selinux/config dynamically depending on build-time configuration options as well as the policy type installed.

Understood, and I agree that the IB should be prioritized over anything else at this point..I believe that the both refpolicy and selinux-policy currently -probably- will allow all systems to boot at least with ramoverlay failover, I am just not a hundred percent sure about all scenario's.

Regardless, I've changed the mode from config back to enforcement as per your suggestion.

aparcar commented:
Please agree on a naming which is consistent for the package. OpenWrt has some [fun](https://github.com/openwrt/openwrt/blob/master/rules.mk#L265) naming hiccups but we shouldn't add more of them.

doverride commented:
> Please agree on a naming which is consistent for the package. OpenWrt has some [fun](https://github.com/openwrt/openwrt/blob/master/rules.mk#L265) naming hiccups but we shouldn't add more of them.

I think "selinux-policy" is fine. In the end its just an identifier.

doverride commented:
Bumped the spec. Added initial support for LuCI. No firmware upgrades package managing, and graphs yet (and more), fixed various bugs. Current version works "nicely" on Linksys m8300 provided that `CONFIG_UBIFS_FS_XATTR` is set. Still awaiting a fix for fstools to do initial labeling of the overlay root, but things should "work" without.

doverride commented:
bumped version again. there was some unfinished business with certificate generation for uhttpd on initial installation, should be addressed now.

doverride commented:
Bumped spec once again. From now on i will be focussing on stabilizing and I wont be adding any new features any time soon. There are two known loose ends: The backup/flash firmware, and software functionality is not yet supported in LuCI. What is supported is LuCI-ACME. So its Base OpenWRT + LuCI + LuCI-ACME. If anyone desires support for more functionality then either fork or contribute. I'll be happy to teach people the ropes.

aparcar commented:
Great thank you for the work, I'll give it a test run today!

flyn-org commented:
This is great, guys! I am working on investigating the modular policy. This might help with things like https://github.com/openwrt/openwrt/pull/3448, but I am not sure yet. I should have some time to run some experiments today. Other distributions have worked through similar challenges; see, for example, https://fedoraproject.org/wiki/SELinux_Policy_Modules_Packaging_Draft.

doverride commented:
Modular policy is just not practical on for example devices with 512MB ram. But other than that, unlike say Fedora, in OpenWrt the policy is loaded from a read-only Squashfs and any changes to the policy at runtime will be disregarded as soon as you reboot.

doverride commented:
Draft build instructions for Fedora:

Ensure you are in ~
```
cd ${HOME}
```
Install build dependencies with sudo
```
sudo dnf install gcc-c++ git make bc make patch wget unzip tar bzip2 gettext ncurses-devel perl-FindBin perl-Data-Dumper perl-Thread-Queue perl-base findutils which diffutils file perl-File-Copy openssl-devel flex libxslt intltool zlib-devel rsync
```
Clone OpenWrt
```
git clone https://github.com/openwrt/openwrt.git
```
Create local package repository lay-out
```
mkdir -p ${HOME}/mypackages/{secilc,selinux-policy/files}
```
Create ${HOME}/mypackages/secilc/Makefile
```
#
# This is free software, licensed under the GNU General Public License v2.
# See /LICENSE for more information.
#

include $(TOPDIR)/rules.mk

PKG_NAME:=secilc
PKG_VERSION:=3.1
PKG_RELEASE:=1

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://github.com/SELinuxProject/selinux/releases/download/20200710
PKG_HASH:=86117246fec3017af710a9ff7c1dae3ed1cd571e232a86cff3e2a3de2d6aa65c
HOST_BUILD_DEPENDS:=libsepol/host

PKG_MAINTAINER:=Dominick Grift <dominick.grift@defensec.nl>
PKG_CPE_ID:=cpe:/a:selinuxproject:secilc
PKG_LICENSE:=BSD-2-Clause
PKG_LICENSE_FILES:=COPYING

include $(INCLUDE_DIR)/host-build.mk
include $(INCLUDE_DIR)/nls.mk
include $(INCLUDE_DIR)/package.mk

HOST_LDFLAGS+=-Wl,-rpath=$(STAGING_DIR_HOSTPKG)/lib
HOST_MAKE_FLAGS += \
	DESTDIR=$(STAGING_DIR_HOSTPKG) \
	PREFIX=

define Package/secilc
	SECTION:=utils
	CATEGORY:=Utilities
	TITLE:=SELinux Common Intermediate Language (CIL) Compiler
	URL:=http://selinuxproject.org/page/Main_Page
endef

define Package/secilc/description
	The SELinux CIL Compiler is a compiler that converts the CIL language as
	described on the CIL design wiki into a kernel binary policy file.
	Please see the CIL Design Wiki at:
	http://github.com/SELinuxProject/cil/wiki/
	for more information about the goals and features on the CIL language.
endef

define Build/Compile
	$(call Build/Compile/Default,secilc)
endef

define Host/Compile
	$(call Host/Compile/Default,secilc)
endef

define Host/Install
	$(INSTALL_DIR) $(STAGING_DIR_HOSTPKG)/bin
	$(INSTALL_BIN) $(HOST_BUILD_DIR)/secilc $(STAGING_DIR_HOSTPKG)/bin
endef

define Package/secilc/install
	$(INSTALL_DIR) $(1)/usr/bin
	$(INSTALL_BIN) $(PKG_BUILD_DIR)/secilc $(1)/usr/bin
endef

$(eval $(call BuildPackage,secilc))
$(eval $(call HostBuild))
```
Create ${HOME}/mypackages/selinux-policy/Makefile
```
#
# This is free software, licensed under the GNU General Public License v2.
# See /LICENSE for more information.
#

include $(TOPDIR)/rules.mk

PKG_NAME:=selinux-policy
PKG_SOURCE_PROTO:=git
PKG_SOURCE_URL:=https://git.defensec.nl/selinux-policy.git
# latest commit date
PKG_SOURCE_DATE:=2020-10-07
# latest commit id
PKG_SOURCE_VERSION:=cdedea73ecbb8968f9736c35c143b586c8371467
PKG_MIRROR_HASH:=skip
PKG_BUILD_DEPENDS:=secilc/host policycoreutils/host

PKG_MAINTAINER:=Dominick Grift <dominick.grift@defensec.nl>
PKG_CPE_ID:=cpe:/a:defensec:selinux-policy
PKG_LICENSE:=Unlicense
PKG_LICENSE_FILES:=LICENSE

include $(INCLUDE_DIR)/package.mk

define Package/selinux-policy
	SECTION:=system
	CATEGORY:=Base system
	TITLE:=SELinux security policy for OpenWrt
	URL:=https://git.defensec.nl/?p=selinux-policy.git;a=summary
	PKGARCH:=all
endef

define Package/selinux-policy/description
	Basic SELinux Security Policy designed specifically for
	OpenWrt and written in Common Intermediate Language.
endef

define Build/Compile
	$(call Build/Compile/Default,policy)
endef

define Package/selinux-policy/conffiles
/etc/selinux/config
endef

define Package/selinux-policy/install
	$(INSTALL_DIR) $(1)/etc/selinux/$(PKG_NAME)/contexts/files/
	$(INSTALL_DIR) $(1)/etc/selinux/$(PKG_NAME)/policy/
	$(INSTALL_DATA) $(PKG_BUILD_DIR)/file_contexts.subs_dist $(1)/etc/selinux/$(PKG_NAME)/contexts/files/
	$(INSTALL_DATA) $(PKG_BUILD_DIR)/file_contexts $(1)/etc/selinux/$(PKG_NAME)/contexts/files/
	$(INSTALL_CONF) $(PKG_BUILD_DIR)/policy.* $(1)/etc/selinux/$(PKG_NAME)/policy/
	$(INSTALL_DATA) ./files/selinux-config $(1)/etc/selinux/config
endef

$(eval $(call BuildPackage,selinux-policy))
```
Create ${HOME}/mypackages/selinux-policy/files/selinux-config
```
SELINUX=permissive
SELINUXTYPE=selinux-policy
```
Add mypackages repository
```
echo "src-link custom ${HOME}/mypackages" >> ${HOME}/openwrt/feeds.conf.default
```
Update the feeds
```
${HOME}/openwrt/scripts/feeds update -a
${HOME}/openwrt/scripts/feeds install -a
```
Change directory into ${HOME}/openwrt
```
cd ${HOME}/openwrt
```
Make menuconfig
```
make -j$((`nproc` + 1)) menuconfig V=99
```
In menuconfig pick a target
```
Target System (Atheros ATH79)
```
In menuconfig pick a subtarget
```
Subtarget (Generic)
```
In menuconfig pick a target profile
```
Target Profile (Default Profile (all drivers))
```
In menuconfig global build settings enable rootfs security labels
```
Global build settings
  [*] Enable rootfs security labels (NEW)
```
In menu config global build settings kernel build options enabel SELinux
```
  Kernel build options
    [*] NSA SELinux Support (NEW)
      [*]   NSA SELinux boot parameter (NEW)
      [*]   NSA SELinux Development Support (NEW)
```
In menuconfig basesystem enable busybox-selinux procd-selinux and selinux-policy (disable plain busybox and procd)
```
Base system
  < > busybox................................ Core utilities for embedded Linux (NEW)
  <*> busybox-selinux... Core utilities for embedded Linux with SELinux support (NEW)
  < > procd..................................... OpenWrt system process manager (NEW)
  <*> procd-selinux........ OpenWrt system process manager with SELinux support (NEW)
  <*> selinux-policy....................... SELinux security policy for OpenWrt
```
In menuconfig basesystem utilities filesystem enable f2fs-tools-selinux
```
  Utilities
    Filesystem
      <*> f2fs-tools-selinux (NEW)
```
Make kernel_menuconfig
```
make -j$((`nproc` + 1)) kernel_menuconfig V=99
```
In kernel menuconfig filesystems miscellaneous filesystems enable ubifs xattr support
```
File systems
  [*] Miscellaneous filesystems
    [*]     UBIFS XATTR support
```
Exclude refpolicy
```
sed -i 's/CONFIG_PACKAGE_refpolicy=y/#CONFIG_PACKAGE_refpolicy=y/' ${HOME}/openwrt/.config
```
Make factory and sysupgrade images
```
make -j$((`nproc` + 1)) V=99
```

Here is a video demonstration of testing and developing with Links WRT1900ACS: https://www.youtube.com/watch?v=WPjOekmsAiA

dangowrt commented:
Pushed to git.openwrt.org, thanks for the great work!
